File: src/commands/request.rs
use diesel::{insert_into, BelongingToDsl, ExpressionMethods, PgConnection, QueryDsl, RunQueryDsl};
use substring::Substring;
use twitch_irc::message::PrivmsgMessage;

use crate::{
    models::diesel::{Channel, ChannelPreference, NewChannel, NewChannelPreference, NewUser, User},
    schema::{channel_preferences::dsl as chp, channels::dsl as ch, users::dsl as us},
    shared_variables::{DEFAULT_LANGUAGE, DEFAULT_PREFIX},
};

use super::CommandLoader;

#[derive(Clone)]
pub struct Request {
    pub command_id: String,
    pub subcommand_id: Option<String>,
    pub message: Option<String>,

    pub sender: User,
    pub channel: Channel,
    pub channel_preference: ChannelPreference,
}

impl Request {
    pub fn try_from(
        message: &PrivmsgMessage,
        command_loader: &CommandLoader,
        conn: &mut PgConnection,
    ) -> Option<Request> {
        let channel = ch::channels
            .filter(ch::alias_id.eq(message.channel_id.parse::<i32>().unwrap()))
            .first::<Channel>(conn)
            .unwrap_or_else(|_| {
                insert_into(ch::channels)
                    .values(vec![NewChannel {
                        alias_id: message.channel_id.parse::<i32>().unwrap(),
                        alias_name: message.channel_login.clone(),
                    }])
                    .execute(conn)
                    .expect("Failed to create a new channel");

                ch::channels
                    .filter(ch::alias_id.eq(message.channel_id.parse::<i32>().unwrap()))
                    .first::<Channel>(conn)
                    .expect("Failed to get a channel after creating it")
            });

        let mut channel_preference = ChannelPreference::belonging_to(&channel)
            .first::<ChannelPreference>(conn)
            .unwrap_or_else(|_| {
                insert_into(chp::channel_preferences)
                    .values(vec![NewChannelPreference {
                        channel_id: channel.id,
                        prefix: DEFAULT_PREFIX.to_string(),
                        language: DEFAULT_LANGUAGE.to_string(),
                    }])
                    .execute(conn)
                    .expect("Failed to create preferences for channel");

                ChannelPreference::belonging_to(&channel)
                    .first::<ChannelPreference>(conn)
                    .expect("Failed to get preferences after creating them")
            });

        let sender = us::users
            .filter(us::alias_id.eq(message.sender.id.parse::<i32>().unwrap()))
            .first::<User>(conn)
            .unwrap_or_else(|_| {
                insert_into(us::users)
                    .values(vec![NewUser {
                        alias_id: message.sender.id.parse::<i32>().unwrap(),
                        alias_name: message.sender.name.clone(),
                    }])
                    .execute(conn)
                    .expect("Failed to create a new user");

                us::users
                    .filter(us::alias_id.eq(message.sender.id.parse::<i32>().unwrap()))
                    .first::<User>(conn)
                    .expect("Failed to get a user after creating it")
            });

        let prefix = channel_preference.prefix.as_str();

        if !message.message_text.starts_with(prefix) {
            return None;
        }

        let message = message
            .message_text
            .substring(prefix.len(), message.message_text.len())
            .to_string();
        let mut message_split = message.split(' ').collect::<Vec<&str>>();

        let command_id = if let Some(word) = message_split.get(0) {
            let word = word.to_string();
            if command_loader
                .commands
                .iter()
                .find(|x| x.get_name().eq(&word))
                .is_some()
            {
                word
            } else {
                return None;
            }
        } else {
            return None;
        };

        let command = command_loader
            .commands
            .iter()
            .find(|x| x.get_name().eq(&command_id))
            .unwrap();

        message_split.remove(0);

        let subcommand_id = if let Some(v) = message_split.get(0) {
            let v = v.to_string();

            if command.get_subcommands().contains(&v) {
                message_split.remove(0);
                Some(v)
            } else {
                None
            }
        } else {
            None
        };

        Some(Request {
            command_id,
            subcommand_id,
            message: if message_split.is_empty() {
                None
            } else {
                Some(message_split.join(" "))
            },
            sender,
            channel,
            channel_preference,
        })
    }
}

File: src/commands/request.cpp
#include "request.hpp"

#include <algorithm>
#include <optional>
#include <string>
#include <vector>

#include "../constants.hpp"

namespace bot {
  namespace command {
    bool Request::fill_request() {
      std::vector<std::string> parts =
          utils::string::split_text(irc_message.message, ' ');

      std::string command_id = parts[0];

      if (command_id.substr(0, DEFAULT_PREFIX.length()) != DEFAULT_PREFIX) {
        return false;
      }

      command_id =
          command_id.substr(DEFAULT_PREFIX.length(), command_id.length());

      bool found = std::any_of(this->command_loader.get_commands().begin(),
                               this->command_loader.get_commands().end(),
                               [&](const auto &command) {
                                 return command->get_name() == command_id;
                               });

      if (!found) {
        return false;
      }

      this->command_id = command_id;

      parts.erase(parts.begin());

      if (parts.empty()) {
        return true;
      }

      std::string subcommand_id = parts[0];
      if (subcommand_id.empty()) {
        this->subcommand_id = std::nullopt;
      } else {
        this->subcommand_id = subcommand_id;
      }
      parts.erase(parts.begin());

      std::string message = utils::string::join_vector(parts, ' ');

      if (message.empty()) {
        this->message = std::nullopt;
      } else {
        this->message = message;
      }

      return true;
    }
  }
}
